File: mip_reference/dotnet-aspnet/Mip/Models/Member.cs
using System.Text.Json.Serialization;

namespace Mip.Models;

/// <summary>
/// Represents a member in the local organization.
/// </summary>
public class Member
{
    public string MemberNumber { get; set; } = "";
    public string? Prefix { get; set; }
    public string FirstName { get; set; } = "";
    public string? MiddleName { get; set; }
    public string LastName { get; set; } = "";
    public string? Suffix { get; set; }
    public string? Honorific { get; set; }
    public string? Rank { get; set; }
    public string? Birthdate { get; set; }
    public int? YearsInGoodStanding { get; set; }
    public string Status { get; set; } = "Active";
    public bool IsActive { get; set; } = true;
    public bool GoodStanding { get; set; } = true;
    public string? Email { get; set; }
    public string? Phone { get; set; }
    public string? Cell { get; set; }
    public Dictionary<string, string?> Address { get; set; } = new();
    public List<Dictionary<string, object?>> Affiliations { get; set; } = new();
    public List<Dictionary<string, object?>> LifeCycleEvents { get; set; } = new();

    [JsonIgnore]
    public string FullName
    {
        get
        {
            var parts = new[] { Prefix, FirstName, MiddleName, LastName, Suffix }
                .Where(p => !string.IsNullOrEmpty(p));
            return string.Join(" ", parts);
        }
    }

    [JsonIgnore]
    public string PartyShortName => $"{FirstName} {LastName?[..1]}";

    [JsonIgnore]
    public string MemberType
    {
        get
        {
            var activeAffiliation = Affiliations.FirstOrDefault(a =>
                a.TryGetValue("is_active", out var active) && active is bool b && b);
            return activeAffiliation?.GetValueOrDefault("member_type")?.ToString() ?? "Member";
        }
    }

    /// <summary>
    /// Format for search response.
    /// </summary>
    public Dictionary<string, object?> ToSearchResult()
    {
        return new Dictionary<string, object?>
        {
            ["member_number"] = MemberNumber,
            ["first_name"] = FirstName,
            ["last_name"] = LastName,
            ["birthdate"] = Birthdate,
            ["contact"] = new Dictionary<string, object?>
            {
                ["email"] = Email,
                ["phone"] = Phone,
                ["address"] = Address
            },
            ["group_status"] = new Dictionary<string, object?>
            {
                ["status"] = Status,
                ["is_active"] = IsActive,
                ["good_standing"] = GoodStanding
            },
            ["affiliations"] = Affiliations.Select(aff => new Dictionary<string, object?>
            {
                ["local_name"] = aff.GetValueOrDefault("local_name"),
                ["local_status"] = aff.GetValueOrDefault("local_status") ?? aff.GetValueOrDefault("status"),
                ["is_active"] = aff.GetValueOrDefault("is_active"),
                ["member_type"] = aff.GetValueOrDefault("member_type")
            }).ToList()
        };
    }

    /// <summary>
    /// Full member profile for COGS.
    /// </summary>
    public Dictionary<string, object?> ToMemberProfile()
    {
        return new Dictionary<string, object?>
        {
            ["member_number"] = MemberNumber,
            ["prefix"] = Prefix,
            ["first_name"] = FirstName,
            ["middle_name"] = MiddleName,
            ["last_name"] = LastName,
            ["suffix"] = Suffix,
            ["honorific"] = Honorific,
            ["rank"] = Rank,
            ["birthdate"] = Birthdate,
            ["years_in_good_standing"] = YearsInGoodStanding,
            ["group_status"] = new Dictionary<string, object?>
            {
                ["status"] = Status,
                ["is_active"] = IsActive
            },
            ["contact"] = new Dictionary<string, object?>
            {
                ["email"] = Email,
                ["phone"] = Phone,
                ["cell"] = Cell,
                ["address"] = Address
            },
            ["affiliations"] = Affiliations,
            ["life_cycle_events"] = LifeCycleEvents
        };
    }

    /// <summary>
    /// Quick status check response.
    /// </summary>
    public Dictionary<string, object?> ToStatusCheck()
    {
        return new Dictionary<string, object?>
        {
            ["member_number"] = MemberNumber,
            ["member_type"] = MemberType,
            ["party_short_name"] = PartyShortName,
            ["group_status"] = new Dictionary<string, object?>
            {
                ["status"] = Status,
                ["is_active"] = IsActive,
                ["good_standing"] = GoodStanding
            }
        };
    }

    /// <summary>
    /// Create from configuration.
    /// </summary>
    public static Member FromConfig(MemberConfig config)
    {
        return new Member
        {
            MemberNumber = config.MemberNumber,
            Prefix = config.Prefix,
            FirstName = config.FirstName,
            MiddleName = config.MiddleName,
            LastName = config.LastName,
            Suffix = config.Suffix,
            Honorific = config.Honorific,
            Rank = config.Rank,
            Birthdate = config.Birthdate,
            YearsInGoodStanding = config.YearsInGoodStanding,
            Status = config.Status,
            IsActive = config.IsActive,
            GoodStanding = config.GoodStanding,
            Email = config.Email,
            Phone = config.Phone,
            Cell = config.Cell,
            Address = config.Address == null ? new() : new Dictionary<string, string?>
            {
                ["address1"] = config.Address.Address1,
                ["address2"] = config.Address.Address2,
                ["city"] = config.Address.City,
                ["state"] = config.Address.State,
                ["postal_code"] = config.Address.PostalCode,
                ["country"] = config.Address.Country
            },
            Affiliations = config.Affiliations.Select(a => new Dictionary<string, object?>
            {
                ["local_name"] = a.LocalName,
                ["local_number"] = a.LocalNumber,
                ["local_location"] = a.LocalLocation,
                ["member_type"] = a.MemberType,
                ["local_status"] = a.LocalStatus,
                ["is_active"] = a.IsActive,
                ["level"] = a.Level,
                ["date"] = a.Date
            }).ToList(),
            LifeCycleEvents = config.LifeCycleEvents.Select(e => new Dictionary<string, object?>
            {
                ["event_name"] = e.EventName,
                ["date"] = e.Date,
                ["local_name"] = e.LocalName
            }).ToList()
        };
    }
}

File: mip_reference/php-symfony/src/Mip/Model/Member.php
<?php

namespace App\Mip\Model;

/**
 * Represents a member in the local organization
 */
class Member
{
    public function __construct(
        public string $memberNumber,
        public string $firstName,
        public string $lastName,
        public ?string $prefix = null,
        public ?string $middleName = null,
        public ?string $suffix = null,
        public ?string $honorific = null,
        public ?string $rank = null,
        public ?string $birthdate = null,
        public ?int $yearsInGoodStanding = null,
        public string $status = 'Active',
        public bool $isActive = true,
        public bool $goodStanding = true,
        public ?string $email = null,
        public ?string $phone = null,
        public ?string $cell = null,
        public array $address = [],
        public array $affiliations = [],
        public array $lifeCycleEvents = []
    ) {}

    public function getFullName(): string
    {
        $parts = array_filter([
            $this->prefix,
            $this->firstName,
            $this->middleName,
            $this->lastName,
            $this->suffix,
        ], fn($p) => $p !== null && $p !== '');

        return implode(' ', $parts);
    }

    public function getPartyShortName(): string
    {
        $lastInitial = !empty($this->lastName) ? $this->lastName[0] : '';
        return "{$this->firstName} {$lastInitial}";
    }

    public function getMemberType(): string
    {
        foreach ($this->affiliations as $affiliation) {
            if ($affiliation['is_active'] ?? false) {
                return $affiliation['member_type'] ?? 'Member';
            }
        }
        return 'Member';
    }

    public function toSearchResult(): array
    {
        return [
            'member_number' => $this->memberNumber,
            'first_name' => $this->firstName,
            'last_name' => $this->lastName,
            'birthdate' => $this->birthdate,
            'contact' => [
                'email' => $this->email,
                'phone' => $this->phone,
                'address' => $this->address,
            ],
            'group_status' => [
                'status' => $this->status,
                'is_active' => $this->isActive,
                'good_standing' => $this->goodStanding,
            ],
            'affiliations' => array_map(fn($aff) => [
                'local_name' => $aff['local_name'] ?? null,
                'local_status' => $aff['local_status'] ?? $aff['status'] ?? null,
                'is_active' => $aff['is_active'] ?? false,
                'member_type' => $aff['member_type'] ?? null,
            ], $this->affiliations),
        ];
    }

    public function toMemberProfile(): array
    {
        return [
            'member_number' => $this->memberNumber,
            'prefix' => $this->prefix,
            'first_name' => $this->firstName,
            'middle_name' => $this->middleName,
            'last_name' => $this->lastName,
            'suffix' => $this->suffix,
            'honorific' => $this->honorific,
            'rank' => $this->rank,
            'birthdate' => $this->birthdate,
            'years_in_good_standing' => $this->yearsInGoodStanding,
            'group_status' => [
                'status' => $this->status,
                'is_active' => $this->isActive,
            ],
            'contact' => [
                'email' => $this->email,
                'phone' => $this->phone,
                'cell' => $this->cell,
                'address' => $this->address,
            ],
            'affiliations' => $this->affiliations,
            'life_cycle_events' => $this->lifeCycleEvents,
        ];
    }

    public function toStatusCheck(): array
    {
        return [
            'member_number' => $this->memberNumber,
            'member_type' => $this->getMemberType(),
            'party_short_name' => $this->getPartyShortName(),
            'group_status' => [
                'status' => $this->status,
                'is_active' => $this->isActive,
                'good_standing' => $this->goodStanding,
            ],
        ];
    }

    public static function fromConfig(array $config): self
    {
        return new self(
            memberNumber: $config['member_number'],
            firstName: $config['first_name'],
            lastName: $config['last_name'],
            prefix: $config['prefix'] ?? null,
            middleName: $config['middle_name'] ?? null,
            suffix: $config['suffix'] ?? null,
            honorific: $config['honorific'] ?? null,
            rank: $config['rank'] ?? null,
            birthdate: $config['birthdate'] ?? null,
            yearsInGoodStanding: $config['years_in_good_standing'] ?? null,
            status: $config['status'] ?? 'Active',
            isActive: $config['is_active'] ?? true,
            goodStanding: $config['good_standing'] ?? true,
            email: $config['email'] ?? null,
            phone: $config['phone'] ?? null,
            cell: $config['cell'] ?? null,
            address: $config['address'] ?? [],
            affiliations: $config['affiliations'] ?? [],
            lifeCycleEvents: $config['life_cycle_events'] ?? []
        );
    }

    public function toArray(): array
    {
        return [
            'member_number' => $this->memberNumber,
            'first_name' => $this->firstName,
            'last_name' => $this->lastName,
            'prefix' => $this->prefix,
            'middle_name' => $this->middleName,
            'suffix' => $this->suffix,
            'honorific' => $this->honorific,
            'rank' => $this->rank,
            'birthdate' => $this->birthdate,
            'years_in_good_standing' => $this->yearsInGoodStanding,
            'status' => $this->status,
            'is_active' => $this->isActive,
            'good_standing' => $this->goodStanding,
            'email' => $this->email,
            'phone' => $this->phone,
            'cell' => $this->cell,
            'address' => $this->address,
            'affiliations' => $this->affiliations,
            'life_cycle_events' => $this->lifeCycleEvents,
        ];
    }

    public static function fromArray(array $data): self
    {
        return self::fromConfig($data);
    }
}
